# Importar Librerias

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Análisis breve de los distintos datasets

In [2]:
car_data = pd.read_csv('../datasets/car data.csv')
car_details = pd.read_csv('../datasets/CAR DETAILS FROM CAR DEKHO.csv')
car_details_v3  = pd.read_csv('../datasets/Car details v3.csv')
car_details_v4  = pd.read_csv('../datasets/Car details v4.csv')

In [3]:
car_data.sample(5)

Car_Name  Year  Selling_Price  Present_Price  Kms_Driven  \
168   Honda CB Trigger  2013           0.42           0.73       12000   
176      Honda Karizma  2011           0.35           1.05       30000   
290              amaze  2014           4.50           6.40       19000   
133  Bajaj Avenger 220  2016           0.72           0.95         500   
268               brio  2017           4.80           5.80       19000   

    Fuel_Type Seller_Type Transmission  Owner  
168    Petrol  Individual       Manual      0  
176    Petrol  Individual       Manual      0  
290    Petrol      Dealer       Manual      0  
133    Petrol  Individual       Manual      0  
268    Petrol      Dealer       Manual      0

In [4]:
car_data.describe(include='all').T

count unique     top freq         mean           std     min  \
Car_Name         301     98    city   26          NaN           NaN     NaN   
Year           301.0    NaN     NaN  NaN  2013.627907      2.891554  2003.0   
Selling_Price  301.0    NaN     NaN  NaN     4.661296      5.082812     0.1   
Present_Price  301.0    NaN     NaN  NaN     7.628472      8.644115    0.32   
Kms_Driven     301.0    NaN     NaN  NaN  36947.20598  38886.883882   500.0   
Fuel_Type        301      3  Petrol  239          NaN           NaN     NaN   
Seller_Type      301      2  Dealer  195          NaN           NaN     NaN   
Transmission     301      2  Manual  261          NaN           NaN     NaN   
Owner          301.0    NaN     NaN  NaN     0.043189      0.247915     0.0   

                   25%      50%      75%       max  
Car_Name           NaN      NaN      NaN       NaN  
Year            2012.0   2014.0   2016.0    2018.0  
Selling_Price      0.9      3.6      6.0      35.0  
Present_Price      1.2      6.4      9.9      92.6  
Kms_Driven     15000.0  32000.0  48767.0  500000.0  
Fuel_Type          NaN      NaN      NaN       NaN  
Seller_Type        NaN      NaN      NaN       NaN  
Transmission       NaN      NaN      NaN       NaN  
Owner              0.0      0.0      0.0       3.0

In [5]:
car_details.sample(5)

name  year  selling_price  km_driven    fuel  \
3496      Mahindra XUV500 W6 2WD  2016        1290000       5000  Diesel   
2583       Chevrolet Beat Diesel  2012         160000     110000  Diesel   
791   Hyundai Xcent 1.2 Kappa SX  2016         350000      12000  Petrol   
2420    Hyundai EON 1.0 Era Plus  2018         100000      21302  Petrol   
4017    Fiat Linea T Jet Emotion  2010         215000      90000  Petrol   

     seller_type transmission         owner  
3496  Individual       Manual   First Owner  
2583  Individual       Manual   First Owner  
791   Individual       Manual   First Owner  
2420  Individual       Manual   First Owner  
4017  Individual       Manual  Second Owner

In [6]:
car_details.describe(include='all').T

count unique                     top  freq           mean  \
name             4340   1491  Maruti Swift Dzire VDI    69            NaN   
year           4340.0    NaN                     NaN   NaN    2013.090783   
selling_price  4340.0    NaN                     NaN   NaN  504127.311751   
km_driven      4340.0    NaN                     NaN   NaN   66215.777419   
fuel             4340      5                  Diesel  2153            NaN   
seller_type      4340      3              Individual  3244            NaN   
transmission     4340      2                  Manual  3892            NaN   
owner            4340      5             First Owner  2832            NaN   

                         std      min        25%       50%       75%  \
name                     NaN      NaN        NaN       NaN       NaN   
year                4.215344   1992.0     2011.0    2014.0    2016.0   
selling_price  578548.736139  20000.0  208749.75  350000.0  600000.0   
km_driven       46644.102194      1.0    35000.0   60000.0   90000.0   
fuel                     NaN      NaN        NaN       NaN       NaN   
seller_type              NaN      NaN        NaN       NaN       NaN   
transmission             NaN      NaN        NaN       NaN       NaN   
owner                    NaN      NaN        NaN       NaN       NaN   

                     max  
name                 NaN  
year              2020.0  
selling_price  8900000.0  
km_driven       806599.0  
fuel                 NaN  
seller_type          NaN  
transmission         NaN  
owner                NaN

In [7]:
car_details_v3.sample(5)

name  year  selling_price  km_driven    fuel  \
4491               Hyundai Accent DLS  2003         108000     200000  Diesel   
6155    Maruti Alto K10 2010-2014 VXI  2010         160000     100000  Petrol   
2200          Maruti Baleno Delta 1.2  2017         530000      30000  Petrol   
7249  Mahindra KUV 100 mFALCON D75 K2  2017         450000     120000  Diesel   
1557         Maruti Wagon R LXI Minor  2009         135000      50000  Petrol   

     seller_type transmission         owner     mileage   engine max_power  \
4491  Individual       Manual  Second Owner   13.1 kmpl  1493 CC    82 bhp   
6155  Individual       Manual   Third Owner  20.92 kmpl   998 CC  67.1 bhp   
2200  Individual       Manual  Second Owner   21.4 kmpl  1197 CC  83.1 bhp   
7249  Individual       Manual   First Owner  25.32 kmpl  1198 CC    77 bhp   
1557  Individual       Manual  Second Owner   18.9 kmpl  1061 CC    67 bhp   

                   torque  seats  
4491       187Nm@ 2500rpm    5.0  
6155        90Nm@ 3500rpm    5.0  
2200       115Nm@ 4000rpm    5.0  
7249  190Nm@ 1750-2250rpm    6.0  
1557        84Nm@ 3500rpm    5.0

In [8]:
car_details_v3.describe(include='all').T

count unique                     top  freq           mean  \
name             8128   2058  Maruti Swift Dzire VDI   129            NaN   
year           8128.0    NaN                     NaN   NaN    2013.804011   
selling_price  8128.0    NaN                     NaN   NaN  638271.807702   
km_driven      8128.0    NaN                     NaN   NaN   69819.510827   
fuel             8128      4                  Diesel  4402            NaN   
seller_type      8128      3              Individual  6766            NaN   
transmission     8128      2                  Manual  7078            NaN   
owner            8128      5             First Owner  5289            NaN   
mileage          7907    393               18.9 kmpl   225            NaN   
engine           7907    121                 1248 CC  1017            NaN   
max_power        7913    322                  74 bhp   377            NaN   
torque           7906    441          190Nm@ 2000rpm   530            NaN   
seats          7907.0    NaN                     NaN   NaN       5.416719   

                         std      min       25%       50%       75%  \
name                     NaN      NaN       NaN       NaN       NaN   
year                4.044249   1983.0    2011.0    2015.0    2017.0   
selling_price  806253.403508  29999.0  254999.0  450000.0  675000.0   
km_driven       56550.554958      1.0   35000.0   60000.0   98000.0   
fuel                     NaN      NaN       NaN       NaN       NaN   
seller_type              NaN      NaN       NaN       NaN       NaN   
transmission             NaN      NaN       NaN       NaN       NaN   
owner                    NaN      NaN       NaN       NaN       NaN   
mileage                  NaN      NaN       NaN       NaN       NaN   
engine                   NaN      NaN       NaN       NaN       NaN   
max_power                NaN      NaN       NaN       NaN       NaN   
torque                   NaN      NaN       NaN       NaN       NaN   
seats               0.959588      2.0       5.0       5.0       5.0   

                      max  
name                  NaN  
year               2020.0  
selling_price  10000000.0  
km_driven       2360457.0  
fuel                  NaN  
seller_type           NaN  
transmission          NaN  
owner                 NaN  
mileage               NaN  
engine                NaN  
max_power             NaN  
torque                NaN  
seats                14.0

In [ ]:
len(car_details_v3['name'].unique())
len(car_details_v3['name'].unique())

In [ ]:
car_details_v4.sample(5)

In [ ]:
car_details_v4.describe(include='all').T

In [ ]:
car_details_v3.columns

| **Car data**      | **CAR DETAILS FROM CAR DEKHO** | **Car details v3** | **Car details v4** |
|-------------------|--------------------------------|--------------------|--------------------|
| 301 observaciones | 4340 observaciones             | 8128 observaciones | 2059 observaciones |
| Car_Name          | name                           | name               | Make + Model       |
| Year              | year                           | year               | Year               |
| Selling_Price     | selling_Price                  | selling_price      | Price              |
| Present Price     | -                              | -                  | -                  |
| Kms_Driven        | km_driven                      | km_driven          | Kilometer          |
| Fuel_Type         | fuel                           | fuel               | Fuel Type          |
| Seller_Type       | seller_type                    | seller_type        | Seller Type        |
| Transmission      | transmission                   | transmission       | Transmission       |
| Owner             | owner                          | owner              | Owner              |
| -                 | -                              | mileage            | -                  |
| -                 | -                              | engine             | Engine             |
| -                 | -                              | max_power          | Max Power          |
| -                 | -                              | torque             | Max Torque         |
| -                 | -                              | seats              | Seating Capacity   |
| -                 | -                              | -                  | Drivetrain         |
| -                 | -                              | -                  | Length             |
| -                 | -                              | -                  | Width              |
| -                 | -                              | -                  | Height             |
| -                 | -                              | -                  | Location           |
| -                 | -                              | -                  | Color              |
| -                 | -                              | -                  | Fuel Tank Capacity |



Luego de revisar superficialmente los datasets, decidimos empezar a trabajar con `Car details v3`. El motivo es que cuenta con la mayor cantidad de observaciones (8128), y además suma 5 features con respecto a `car data` y `CAR DETAILS FROM CAR DEKHO`, que nos parece pueden resultar reelevantes en el entrenamiento de un modelo.

`Car details v4`, aunque con solo 2059 observaciones, cuenta con otras 7 features, algunas que pueden asumirse como interesantes para calcular el precio. Además, este dataset cuenta con valores de autos más actualizados que el dataset v3, por lo tanto podría suponerse que genere un mejor modelo para predecir el valor de autos de los últimos años. Eventualmente vamos a intentar explorar un poco las posibilidades con este dataset también, aunque manteniendo mayor foco en el dataset v3.

In [ ]:
for column in car_details_v3.columns:
    unique_values = car_details_v3[column].unique()
    unique_count = len(unique_values)
    print(f"Feature: {column}")
    print(f"Number of unique values: {unique_count}")
    print(f"Unique values: {unique_values}\n")


In [ ]:
car_details_v3.describe(include='all').T

In [ ]:
print(car_details_v3['fuel'].unique())
print(car_details_v3['seller_type'].unique())
print(car_details_v3['transmission'].unique())
print(car_details_v3['owner'].unique())
print(car_details_v3['year'].agg({ min: 'min', max: 'max'}))

### Tipos de datos:

- **Variables categóricas**: `name`, `fuel`, `seller_type`, `transmission`, `owner`, `engine`, `max_power`, `torque`, `seats`.
- **Variables de tipo fecha**: `year`.
- **Variables numéricas**: `km_driven`, `mileage`.
- **Target numérico**: `selling_price`

### Análisis de los datos:

- `name`, `engine`, `max_power` y `torque` tienen alta cardinalidad (2058, 121, 322 y 441 posibles valores, respectivamente).
- `fuel` tiene 4 valores posibles: `Diesel`, `Petrol`, `LPG` y `CNG`.
- `seller_type` tiene 3 valores posibles: `Individual`, `Dealer` y `Trustmark Dealer`.
- `transmission` puede ser `Manual` o `Automatic`.
- `owner` tiene 5 valores posibles: `First Owner`, `Second Owner`, `Third Owner`, `Fourth & Above Owner` y `Test Drive Car`
- Contamos con observaciones desde el año `1983` al `2020`.

Veamos distribución de las observaciones según algunas de las features:

In [ ]:
transmission_counts = car_details_v3['transmission'].value_counts()

plt.figure(figsize=(8, 8))
plt.pie(
    transmission_counts,
    labels=transmission_counts.index,
    autopct='%1.1f%%',
    startangle=140,
    textprops={'color': 'black'},
    wedgeprops={'edgecolor': 'black'}
)
plt.title('Distribución de Tipo de Transmisión')
plt.show()

In [ ]:
seller_type_counts = car_details_v3['seller_type'].value_counts()

plt.figure(figsize=(8, 8))
plt.pie(
    seller_type_counts,
    labels=seller_type_counts.index,
    autopct='%1.1f%%',
    startangle=140,
    textprops={'color': 'black'},
    wedgeprops={'edgecolor': 'black'}
)
plt.title('Distribución por Tipo de Vendedor')
plt.show()

In [ ]:
fuel_counts = car_details_v3['fuel'].value_counts()

plt.figure(figsize=(8, 8))
plt.pie(
    fuel_counts,
    labels=fuel_counts.index,
    autopct='%1.1f%%',
    startangle=140,
    textprops={'color': 'black'},
    wedgeprops={'edgecolor': 'black'}
)
plt.title('Distribución por Tipo de Combustible')
plt.show()

In [ ]:
owner_counts = car_details_v3['owner'].value_counts()

plt.figure(figsize=(8, 8))
plt.pie(
    owner_counts,
    labels=owner_counts.index,
    autopct='%1.1f%%',
    startangle=140,
    textprops={'color': 'black'},
    wedgeprops={'edgecolor': 'black'}
)
plt.title('Distribución por Cantidad de Dueños')
plt.show()

In [ ]:
missing_values = car_details_v3.isnull().sum()

for column, missing in missing_values.items():
    print(f"Feature: {column} - Missing values: {missing}")

Vemos que en las 5 features donde tenemos valores faltantes, los mismos responden a aproximadamente un 2,7% de las observaciones.

In [ ]:
missing_all = car_details_v3[
    car_details_v3[['mileage', 'engine', 'max_power', 'torque', 'seats']].isnull().all(axis=1)
]

rows_missing_all_features = missing_all.shape[0]
print(f"Cantidad de filas con todas las características 'mileage', 'engine', 'max_power', 'torque' y 'seats' faltantes: {rows_missing_all_features}")

In [ ]:
missing_all = car_details_v3[
    car_details_v3[['mileage', 'engine', 'torque', 'seats']].isnull().all(axis=1)
]

rows_missing_all_features = missing_all.shape[0]
print(f"Cantidad de filas con todas las características 'mileage', 'engine', 'torque' y 'seats' faltantes: {rows_missing_all_features}")

In [ ]:
transmission_counts

In [ ]:
seller_type_counts

In [ ]:
owner_counts

In [ ]:
fuel_counts

In [ ]:
year_counts = car_details_v3['year'].value_counts().sort_index()

plt.figure(figsize=(10, 6))
plt.bar(year_counts.index, year_counts.values, color='skyblue')
plt.xlabel('Año')
plt.ylabel('Cantidad de observaciones')
plt.title('Cantidad de observaciones por Año')
plt.xticks(rotation=45)
plt.show()

In [ ]:
year_counts

Parece una distribución sesgada a izquierda.

In [ ]:
car_details_v3['km_driven'].plot(
    kind='box',
    figsize=(10, 10)
)

min_km_driven = car_details_v3['km_driven'].min()
max_km_driven = car_details_v3['km_driven'].max()

print(f"Valor mínimo de km_driven: {min_km_driven}")
print(f"Valor máximo de km_driven: {max_km_driven}")

Todos los valores parecen lógicos. Veamos los que tienen menos de 10000 km

In [ ]:
car_details_v3[car_details_v3['km_driven'] < 10000].sort_values(by='km_driven', ascending=True)

In [ ]:
mileage_non_null = car_details_v3[car_details_v3['mileage'].notnull()].copy()

# Limpiar y convertir 'mileage' a numérico
mileage_non_null['mileage_as_number'] = mileage_non_null['mileage'].str.replace(r'[^0-9.]+', '', regex=True)
mileage_non_null['mileage_as_number'] = pd.to_numeric(mileage_non_null['mileage_as_number'], errors='coerce')

# Crear el boxplot para 'mileage_as_number'
mileage_non_null['mileage_as_number'].plot(
    kind='box',
    figsize=(10, 10)
)
plt.title("Boxplot de 'mileage' (Solo filas con valores)")
plt.ylabel('Mileage')
plt.show()


In [ ]:
mileage_non_null[mileage_non_null['mileage_as_number'] < 10].sort_values(by='mileage_as_number', ascending=True)

In [ ]:
car_details_v3[car_details_v3['seats'] < 2].sort_values(by='seats', ascending=True)


In [ ]:
car_details_v3.sample(50)

In [ ]:
def impute_nulls_with_similar(df):
    similar_found = 0
    non_similar_found = 0
    # Asegurarse de que 'name' y 'year' no tengan valores nulos
    df['name'] = df['name'].fillna('Unknown')
    df['year'] = df['year'].fillna(df['year'].mode()[0])
    
    # Crear un filtro para encontrar filas con valores nulos en las columnas a imputar
    null_condition = df[['mileage', 'engine', 'max_power', 'torque', 'seats']].isnull().any(axis=1)
    
    # Iterar sobre las filas que cumplen con la condición
    for idx, row in df[null_condition].iterrows():
        # Buscar vehículos similares (mismo nombre y año)
        similar_vehicles = df[(df['name'] == row['name']) & 
                              (df['year'] == row['year']) & 
                              (df[['mileage', 'engine', 'max_power', 'torque', 'seats']].notnull().any(axis=1))]
        
        if not similar_vehicles.empty:
            # Si se encuentran vehículos similares, usar el valor más cercano
            for column in ['mileage', 'engine', 'max_power', 'torque', 'seats']:
                if pd.isnull(row[column]):
                    a = 1
                    # df.at[idx, column] = similar_vehicles[column].iloc[0]
            similar_found = similar_found + 1
        else:
            # Si no se encuentran vehículos similares, usar la moda de la columna
            print(f"Imputing for {row['year']}-{row['name']} without similarities")
            for column in ['mileage', 'engine', 'max_power', 'torque', 'seats']:
                if pd.isnull(row[column]):
                    a = 1
                    # df.at[idx, column] = df[column].mode()[0]
            non_similar_found = non_similar_found + 1

                    
    print(f"Imputing {similar_found} vehicles with similarities")
    print(f"Imputing {non_similar_found} vehicles with non similarities")
    return df

def impute_invalid_with_similar(df):
    similar_found = 0
    non_similar_found = 0
    # Asegurarse de que 'name' y 'year' no tengan valores nulos
    df['name'] = df['name'].fillna('Unknown')
    df['year'] = df['year'].fillna(df['year'].mode()[0])
    
    # Iterar sobre las filas con valores nulos en la columna actual
    for idx, row in df[(df['mileage'] == '0.0 kmpl') | (df['max_power'].isin(['bhp', 0]))].iterrows():
        # Buscar vehículos similares (mismo nombre y año)
        similar_vehicles = df[(df['name'] == row['name']) & 
                              (df['year'] == row['year']) & 
                              (df[column].notnull())]

        print(f"Imputing for {row['year']}-{row['name']} without similarities")
        if not similar_vehicles.empty:
            # Si se encuentran vehículos similares, usar el valor más cercano
            # df.at[idx, column] = similar_vehicles[column].iloc[0]
            similar_found = similar_found + 1
        else:
            # Si no se encuentran vehículos similares, usar la moda de la columna
            # df.at[idx, column] = df[column].mode()[0]
            non_similar_found = non_similar_found + 1
    
    print(f"Imputing {similar_found} vehicles with similarities")
    print(f"Imputing {non_similar_found} vehicles with non similarities")
    return df

In [ ]:
car_details_v3  = pd.read_csv('../datasets/Car details v3.csv')

In [ ]:
car_details_v3[car_details_v3['name'] == 'Maruti Swift 1.3 ZXI']

In [ ]:
car_details_v3 = impute_nulls_with_similar(car_details_v3)

In [ ]:
car_details_v3

In [ ]:
def impute_invalid_with_similar(df):
    similar_found = 0
    non_similar_found = 0
    # Asegurarse de que 'name' y 'year' no tengan valores nulos
    df['name'] = df['name'].fillna('Unknown')
    df['year'] = df['year'].fillna(df['year'].mode()[0])
    
    # Iterar sobre las filas con valores nulos en la columna actual
    for idx, row in df[(df['mileage'] == '0.0 kmpl') | (df['max_power'].isin(['bhp', 0]))].iterrows():
        # Buscar vehículos similares (mismo nombre y año)
        similar_vehicles = df[(df['name'] == row['name']) & 
                              (df['year'] == row['year']) & 
                              (df[column].notnull())]
        
        if not similar_vehicles.empty:
            # Si se encuentran vehículos similares, usar el valor más cercano
            df.at[idx, column] = similar_vehicles[column].iloc[0]
            similar_found = similar_found + 1
        else:
            non_similar_found = non_similar_found + 1
            # Si no se encuentran vehículos similares, usar la moda de la columna
            # df.at[idx, column] = df[column].mode()[0]

    print(f"Imputing {similar_found} vehicles with similarities")
    print(f"Imputing {non_similar_found} vehicles with non similarities")
    return df

In [ ]:
car_details_v3 = pd.read_csv('../datasets/Car details v3.csv')

In [ ]:
car_details_v3 = impute_invalid_with_similar(car_details_v3)

In [ ]:
mileage_non_null = car_details_v3[car_details_v3['mileage'].notnull()].copy()

# Limpiar y convertir 'mileage' a numérico
car_details_v3['mileage_as_number'] = mileage_non_null['mileage'].str.replace(r'[^0-9.]+', '', regex=True)
car_details_v3['mileage_as_number'] = pd.to_numeric(car_details_v3['mileage_as_number'], errors='coerce')

# Limpiar y convertir 'engine' a numérico
car_details_v3['engine_as_number'] = mileage_non_null['engine'].str.replace(r'[^0-9.]+', '', regex=True)
car_details_v3['engine_as_number'] = pd.to_numeric(car_details_v3['engine_as_number'], errors='coerce')

# Limpiar y convertir 'max_power' a numérico
car_details_v3['max_power_as_number'] = mileage_non_null['max_power'].str.replace(r'[^0-9.]+', '', regex=True)
car_details_v3['max_power_as_number'] = pd.to_numeric(car_details_v3['max_power_as_number'], errors='coerce')

car_details_v3

In [ ]:
import re

def standardize_torque(torque_str):
    if pd.isna(torque_str):
        return np.nan

    # Match the torque value, unit, and RPM value
    torque_match = re.findall(r"(\d*\.?\d+)\s*(kgm|Nm|nm)?\s*@\s*([-\d\s,]*)\s*(rpm)?", torque_str)

    # If a match is found, process the first match
    if torque_match:
        value, unit, rpm_range, _ = torque_match[0]
        value = float(value)

        # Check for kgm and convert to Nm
        if 'kgm' in torque_str:
            value = value * 9.81  # Convert kgm to Nm

        # Format torque value to 2 decimal places
        torque_value = f"{value:.2f}Nm"

        # Clean and format the rpm_range
        rpm_value = rpm_range.strip().replace(',', '')  # Remove commas and whitespace

        # Combine torque and RPM value
        if rpm_value:
            return f"{torque_value} {rpm_value}rpm"
        else:
            return torque_value  # Return only the torque value if no RPM is found

    # Alternative regex to check for kgm in parentheses
    torque_match_alt = re.findall(r"(\d*\.?\d+)\s*@\s*([-\d\s,]*)\s*\(kgm@\s*rpm\)", torque_str)

    if torque_match_alt:
        value, rpm_range = torque_match_alt[0]
        value = float(value) * 9.81  # Convert kgm to Nm

        # Format torque value to 2 decimal places
        torque_value = f"{value:.2f}Nm"
        rpm_value = rpm_range.strip().replace(',', '')  # Remove commas and whitespace

        return f"{torque_value} {rpm_value}rpm"

    # Check for alternative formats (e.g., kgm without @)
    torque_match_kgm = re.findall(r"(\d*\.?\d+)\s*kgm\s*at\s*([-\d\s,]*)\s*(rpm)?", torque_str)

    if torque_match_kgm:
        value, rpm_range, _ = torque_match_kgm[0]
        value = float(value) * 9.81  # Convert kgm to Nm

        # Format torque value to 2 decimal places
        torque_value = f"{value:.2f}Nm"
        rpm_value = rpm_range.strip().replace(',', '')  # Remove commas and whitespace

        # Combine torque and RPM value
        if rpm_value:
            return f"{torque_value} {rpm_value}rpm"
        else:
            return torque_value  # Return only the torque value if no RPM is found

    # If no valid torque value is found, return NaN
    return np.nan


In [ ]:
car_details_v3['torque_standardized'] = car_details_v3['torque'].apply(standardize_torque)
car_details_v3['torque_standardized'] = car_details_v3['torque_standardized'].round(2)

car_details_v3

In [ ]:
import pandas as pd

def impute_nulls_with_similar(df):
  # Ensure 'name' and 'year' have non-null values
  df['name'] = df['name'].fillna('Unknown')
  df['year'] = df['year'].fillna(df['year'].mode()[0])

  # Columns to impute
  impute_cols = ['mileage', 'engine', 'max_power', 'torque', 'seats']

  # Create a filtration mask for missing values
  null_mask = df[impute_cols].isnull().any(axis=1)
  # Create a new column to indicate imputation
  df['imputed'] = 0

  for idx, row in df[null_mask].iterrows():
    # Find similar vehicles based on name, year, and fuel (if fuel is not null)
    similar_condition = (df['name'] == row['name']) & (df['year'] == row['year']) & (df['fuel'] == row['fuel'])
    similar_vehicles = df[similar_condition]

    if not similar_vehicles.empty:
      # Impute with the closest vehicle's value for missing columns
      for col in impute_cols:
        if pd.isnull(row[col]):
          df.at[idx, col] = similar_vehicles[col].iloc[0]
          df.at[idx, 'imputed'] = 1

  return df

In [ ]:
car_details_v3 = impute_nulls_with_similar(car_details_v3)

car_details_v3

In [ ]:
# car_details_v3[['make', 'model']] = car_details_v3['name'].str.split(' ', n=1, expand=True)

# car_details_v3['Model'].unique()

In [ ]:
# Find rows where either 'Make' or 'Model' is empty
# empty_make_model = car_details_v3[(car_details_v3['Make'] == '') | (car_details_v3['Model'] == '')]

# Print the rows with empty values
# empty_make_model

In [ ]:
from sklearn.feature_extraction import FeatureHasher
from sklearn.impute import KNNImputer

def hashing_encoding(df, cols, data_percent=0.85, verbose=False):
    for i in cols:
        val_counts = df[i].value_counts(dropna=False)
        s = sum(val_counts.values)
        h = val_counts.values / s
        c_sum = np.cumsum(h)
        c_sum = pd.Series(c_sum)
        n = c_sum[c_sum > data_percent].index[0]
        if verbose:
            print("n hashing para ", i, ":", n)
        if n > 0:
            fh = FeatureHasher(n_features=n, input_type='string')
            hashed_features = fh.fit_transform(
                df[i].astype(str).values.reshape(-1, 1)
            ).todense()
            df = df.join(pd.DataFrame(hashed_features).add_prefix(i + '_'))

    return df.drop(columns=cols)

def knn_imputer(df):

    cat_cols = ['name', 'fuel', 'owner', 'seller_type', 'transmission']

    # Aplicamos hashing para las categoricas
    df = hashing_encoding(df, cat_cols)

    # Eliminamos name y alignment para imputar
    df = df.drop(columns=['selling_price', 'mileage', 'engine', 'max_power', 'torque', 'seats', 'torque_standardized'])

    # definimos un n arbitrario
    imputer = KNNImputer(n_neighbors=2, weights="uniform")
    df = pd.DataFrame(imputer.fit_transform(df), columns=df.columns)
    return df


knn_imputation = knn_imputer(car_details_v3).add_suffix('_knn')
knn_imputation

car_details_v3 = knn_imputation['']

In [ ]:
import pandas as pd
from sklearn.impute import KNNImputer

# Optional: Encode categorical features
car_details_v3 = pd.get_dummies(car_details_v3, columns=['fuel', 'name', 'transmission', 'seller_type', 'owner', 'torque_standardized'], drop_first=True)

car_details_v3 = car_details_v3.drop(columns=['mileage', 'engine', 'max_power', 'torque'])

In [ ]:
# Initialize the KNNImputer
imputer = KNNImputer(n_neighbors=5)

# Create indicator columns for missing values
for column in ['mileage_as_number', 'engine_as_number', 'max_power_as_number', 'seats', 'torque_standardized']:
    car_details_v3[f'{column}_was_imputed'] = car_details_v3[column].isnull().astype(int)

# Impute missing values for engine, torque, and seats
columns_to_impute = ['mileage_as_number', 'engine_as_number', 'max_power_as_number', 'seats', 'torque_standardized']
car_details_v3[columns_to_impute] = imputer.fit_transform(car_details_v3[columns_to_impute])

# Set the indicator to 1 for the rows where values were imputed
for column in ['mileage_as_number', 'engine_as_number', 'max_power_as_number', 'seats', 'torque_standardized']:
    car_details_v3[f'{column}_was_imputed'] = car_details_v3[f'{column}_was_imputed'] | car_details_v3[column].isnull().astype(int)

In [ ]:
car_details_v3

In [ ]:
numerical_data = car_details_v3[['year', 'selling_price', 'km_driven', 'seats', 'mileage_as_number', 'engine_as_number', 'max_power_as_number']]

# Calcular la matriz de correlación
correlation_matrix = numerical_data.corr()

# Configurar el tamaño del gráfico
plt.figure(figsize=(10, 8))

# Crear el mapa de calor
sns.heatmap(correlation_matrix, annot=True, fmt=".2f", cmap='coolwarm', square=True, cbar_kws={"shrink": .8})

# Añadir título
plt.title('Matriz de Correlación para Columnas Numéricas')

# Mostrar el gráfico
plt.show()